File: Examples/VYNFCKitExampleObjc/VYNFCKitExampleObjc/ViewController.m
//
//  ViewController.m
//  VYNFCKitExampleObjc
//
//  Created by Vince Yuan on 7/11/17.
//  Copyright © 2017 Vince Yuan. All rights reserved.
//

#import "ViewController.h"
#import <CoreNFC/CoreNFC.h>
#import <VYNFCKit/VYNFCKit.h>

@interface ViewController () <NFCNDEFReaderSessionDelegate> {
    NSString *_results;
    __weak IBOutlet UITextView *_textViewResults;
    __weak IBOutlet UISwitch *_switchOpenURI;
}

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)didTapScanNFCTagButton:(id)sender {
    _results = @"";
    _textViewResults.text = @"";
    NFCNDEFReaderSession *session = [[NFCNDEFReaderSession alloc] initWithDelegate:self queue:dispatch_get_main_queue() invalidateAfterFirstRead:NO];
    [session beginSession];
}

#pragma mark - NFCNDEFReaderSessionDelegate

- (void)readerSession:(nonnull NFCNDEFReaderSession *)session didDetectNDEFs:(nonnull NSArray<NFCNDEFMessage *> *)messages {
    for (NFCNDEFMessage *message in messages) {
        for (NFCNDEFPayload *payload in message.records) {
            id parsedPayload = [VYNFCNDEFPayloadParser parse:payload];
            if (parsedPayload) {
                NSString *text = @"";
                NSString *urlString = nil;
                if ([parsedPayload isKindOfClass:[VYNFCNDEFTextPayload class]]) {
                    text = @"[Text payload]\n";
                    text = [NSString stringWithFormat:@"%@%@", text, ((VYNFCNDEFTextPayload *)parsedPayload).text];
                } else if ([parsedPayload isKindOfClass:[VYNFCNDEFURIPayload class]]) {
                    text = @"[URI payload]\n";
                    text = [NSString stringWithFormat:@"%@%@", text, ((VYNFCNDEFURIPayload *)parsedPayload).URIString];
                    urlString = text;
                } else if ([parsedPayload isKindOfClass:[VYNFCNDEFTextXVCardPayload class]]) {
                    text = @"[TextXVCard payload\n]";
                    text = [NSString stringWithFormat:@"%@%@", text, ((VYNFCNDEFTextXVCardPayload *)parsedPayload).text];
                } else if ([parsedPayload isKindOfClass:[VYNFCNDEFSmartPosterPayload class]]) {
                    text = @"[SmartPoster payload]\n";
                    VYNFCNDEFSmartPosterPayload *sp = parsedPayload;
                    for (VYNFCNDEFTextPayload *textPayload in sp.payloadTexts) {
                        text = [NSString stringWithFormat:@"%@%@\n", text, textPayload.text];
                    }
                    text = [NSString stringWithFormat:@"%@%@", text, sp.payloadURI.URIString];
                    urlString = sp.payloadURI.URIString;
                } else if ([parsedPayload isKindOfClass:[VYNFCNDEFWifiSimpleConfigPayload class]]) {
                    text = @"[WifiSimpleConfig payload]\n";
                    VYNFCNDEFWifiSimpleConfigPayload *wifi = parsedPayload;
                    text = [NSString stringWithFormat:@"%@SSID: %@\nPassword: %@\nMac Address: %@\nAuth Type: %@\nEncrypt Type: %@",
                            text, wifi.credential.ssid, wifi.credential.networkKey, wifi.credential.macAddress,
                            [VYNFCNDEFWifiSimpleConfigCredential authTypeString:wifi.credential.authType],
                            [VYNFCNDEFWifiSimpleConfigCredential encryptTypeString:wifi.credential.encryptType]];
                    if (wifi.version2) {
                        text = [NSString stringWithFormat:@"%@\nVersion2: %@",
                                text, wifi.version2.version];
                    }
                } else {
                    text = @"Parsed but unhandled payload type";
                }
                NSLog(@"%@", text);
                _results = [NSString stringWithFormat:@"%@%@\n", _results, text];
                dispatch_async(dispatch_get_main_queue(), ^{
                    _textViewResults.text = _results;

                    if (_switchOpenURI.isOn && urlString) {
                        // Close NFC reader session and open URI after delay. Not all can be opened on iOS.
                        [session invalidateSession];
                        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                            [[UIApplication sharedApplication] openURL:[NSURL URLWithString: urlString] options:@{} completionHandler:nil];
                        });
                    }
                });
            }
        }
    }
}

- (void)readerSession:(nonnull NFCNDEFReaderSession *)session didInvalidateWithError:(nonnull NSError *)error {
    NSLog(@"%@", error);
    _results = [NSString stringWithFormat:@"%@%@\n", _results, error];
    dispatch_async(dispatch_get_main_queue(), ^{
        _textViewResults.text = _results;
    });
}



@end

File: Examples/VYNFCKitExampleSwift/VYNFCKitExampleSwift/ViewController.swift
//
//  ViewController.swift
//  VYNFCKitExampleSwift
//
//  Created by Vince Yuan on 8/26/17.
//  Copyright © 2017 Vince Yuan. All rights reserved.
//

import UIKit
import CoreNFC
import VYNFCKit

class ViewController: UIViewController, NFCNDEFReaderSessionDelegate {

    var results: String = ""
    @IBOutlet weak var textViewResults: UITextView!
    @IBOutlet weak var switchOpenURI: UISwitch!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func didTapScanNFCTagButton(_ sender: Any) {
        results = ""
        textViewResults.text = ""
        let session = NFCNDEFReaderSession(delegate: self, queue: DispatchQueue.main, invalidateAfterFirstRead: false)
        session.begin()
    }

    func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        for message in messages {
            for payload in message.records {
                guard let parsedPayload = VYNFCNDEFPayloadParser.parse(payload) else {
                    continue
                }
                var text = ""
                var urlString = ""
                if let parsedPayload = parsedPayload as? VYNFCNDEFTextPayload {
                    text = "[Text payload]\n"
                    text = String(format: "%@%@", text, parsedPayload.text)
                } else if let parsedPayload = parsedPayload as? VYNFCNDEFURIPayload {
                    text = "[URI payload]\n"
                    text = String(format: "%@%@", text, parsedPayload.uriString)
                    urlString = text
                } else if let parsedPayload = parsedPayload as? VYNFCNDEFTextXVCardPayload {
                    text = "[TextXVCard payload]\n"
                    text = String(format: "%@%@", text, parsedPayload.text)
                } else if let sp = parsedPayload as? VYNFCNDEFSmartPosterPayload {
                    text = "[SmartPoster payload]\n"
                    for textPayload in sp.payloadTexts {
                        if let textPayload = textPayload as? VYNFCNDEFTextPayload {
                            text = String(format: "%@%@\n", text, textPayload.text)
                        }
                    }
                    text = String(format: "%@%@", text, sp.payloadURI.uriString)
                    urlString = sp.payloadURI.uriString
                } else if let wifi = parsedPayload as? VYNFCNDEFWifiSimpleConfigPayload {
                    text = String(format: "%@SSID: %@\nPassword: %@\nMac Address: %@\nAuth Type: %@\nEncrypt Type: %@",
                                  text, wifi.credential.ssid, wifi.credential.networkKey, wifi.credential.macAddress,
                                  VYNFCNDEFWifiSimpleConfigCredential.authTypeString(wifi.credential.authType),
                                  VYNFCNDEFWifiSimpleConfigCredential.encryptTypeString(wifi.credential.encryptType)
                    )
                    if let version2 = wifi.version2 {
                        text = String(format: "%@\nVersion2: %@", text, version2.version)
                    }
                } else {
                    text = "Parsed but unhandled payload type"
                }
                print(text)

            }
        }
    }

    func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        print(error)
    }

}

